pkg/observable/channel/observable_test.go

```diff
@@ -0,0 +1,366 @@
+		},
+		{
+			name:            "empty non-buffered producer",
+			producer:        make(chan *int),
+			inputs:          inputs,
+			expectedOutputs: inputs,
+		},
+		{
+			name:            "empty buffered len 1 producer",
+			producer:        make(chan *int, 1),
+			inputs:          inputs,
+			expectedOutputs: inputs,
+		},
+		// INCOMPLETE: producer channels which are full are proving harder to test
+		// robustly (no flakiness); perhaps it has to do with the lack of some
+		// kind of guarantee about the receiver order on the consumer side.
+		//
+		// The following scenarios should generally pass but are flaky:
+		//
+		// {
+		// 	name:            "full non-buffered producer",
+		// 	producer:        fullBlockingProducer,
+		// 	inputs:          inputs[1:],
+		// 	expectedOutputs: inputs,
+		// 	setupFn: func(t test) {
+		// 		go func() {
+		// 			// blocking send
+		// 			t.producer <- &inputs[0]
+		// 		}()
+		// 	},
+		// },
+		// {
+		// 	name:            "full buffered len 1 producer",
+		// 	producer:        fullBufferedProducer,
+		// 	inputs:          inputs[1:],
+		// 	expectedOutputs: inputs,
+		// 	setupFn: func(t test) {
+		// 		// non-blocking send
+		// 		t.producer <- &inputs[0]
+		// 	},
+		// },
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.setupFn != nil {
+				tt.setupFn(tt)
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			t.Cleanup(cancel)
+
+			obsvbl, producer := channel.NewObservable[*int](
+				channel.WithProducer(tt.producer),
+			)
+			require.NotNil(t, obsvbl)
+			require.NotNil(t, producer)
+			produce := produceWithDelay(producer, productionDelay)
+
+			// construct 3 distinct observers, each with its own channel
+			observers := make([]observable.Observer[*int], 1)
+			for i := range observers {
+				observers[i] = obsvbl.Subscribe(ctx)
+			}
+
+			group, ctx := errgroup.WithContext(ctx)
+
+			// ensure all obsvr channels are notified
+			for obsvrIdx, obsvr := range observers {
+				next := func(outputIndex int, output *int) error {
+					// obsvr channel should receive notified input
+					if !assert.Equalf(
+						t, tt.expectedOutputs[outputIndex],
+						*output,
+						"obsvr Idx: %d", obsvrIdx,
+					) {
+						return fmt.Errorf("unexpected output")
+					}
+					return nil
+				}
+
+				done := func(outputs []*int) error {
+					if !assert.Equalf(
+						t, len(tt.expectedOutputs),
+						len(outputs),
+						"obsvr addr: %p", obsvr,
+					) {
+						return fmt.Errorf("unexpected number of outputs")
+					}
+					return nil
+				}
+
+				// concurrently await notification or timeout to avoid blocking on
+				// empty and/or non-buffered producers.
+				group.Go(goNotifiedOrTimedOutFactory(obsvr, next, done, notifyTimeout))
+			}
+
+			// notify with test input
+			for _, input := range tt.inputs {
+				inputPtr := new(int)
+				*inputPtr = input
+
+				// simulating IO delay in sequential message production
+				produce(inputPtr)
```

Comment:
Where is this function defined?

Reply:
![image](https://github.com/pokt-network/poktroll/assets/600733/06088153-3520-46c3-98c1-2ae30cb8e52e)

Reply:
I've renamed and moved it to be closer.